Object Serialization and Attribute Embedding

faultyserver opened the pull request:
Dealing with #5 (finally) and #11.

Object Serialization takes place implicitly when the WAMP client in Transport (provided by [ruby_wamp_client](https://github.com/ericchapman/ruby_wamp_client)) calls `to_json` on args and kwargs.

Serialization is implemented as a module (`ObjectSerialization`) that `Shark::Object` includes. `Shark::Object` also now includes `Configurable`, and there are a number of options that can be changed to affect the serialization (see `config/serialization.rb` for details).

faultyserver committed:
"vehilce" was spelled wrong in `ve_embed_objects!`

faultyserver committed:
(#11) Start work on ObjectEmbed middleware
Includes the definition of the ObjectEmbed class, as well as a configuration file and additions to the citybus configuration to use the middleware.

faultyserver committed:
ObjectEmbed comes *after* Conductor in the middleware stack
See #11. Having ObjectEmbed come after Conductor ensures that objects are embedded appropriately in all events that come out from the system.

Since Conductor is responsible for publishing new events based on existing ones, if ObjectEmbed were to come before it, then those new events would either not have embedded objects, or Conductor would have to embed them manually (neither of which is a good solution).

faultyserver committed:
Conductor's event handling system is now the default.
The pattern of running unique callbacks for different event types is useful in a lot of middlewares.

For example, ObjectEmbed will likely use the same event handling system to determine which objects should be embedded in an argument list.

faultyserver committed:
`@@event_handlers` should be a class instance variable
As it was (a class variable), event handlers would be defined on the Shark::Middleware class itself, meaning all middlewares would share the same set of event handlers, which defeats the entire purpose of having middleware instances.

As it is now (a class instance variable), each middleware instance gets it's own set of event handlers.

faultyserver committed:
Add `Shark::Object#to_json`
This is a first step in getting middlewares to pass full Objects, rather than simple hashes. Assuming "ruby_wamp_client" uses `to_json` to serialize data (which I think it would, but I can't find any documentation), passing an array of objects as args should just work.

This avoids the need to write complex/ugly serialization logic that checks the types of objects in `args` and conditionally calls `to_h` on them if they are Object instances.

faultyserver committed:
Continue/fix 81224e0. `Middleware#call` assumed incorrect scoping
It was using `@event_handlers` as an instance variable, when it should be accessed as a class instance variable. The remedy is creating the method `Middleware.event_handlers`, which returns the class instance variable and initializes it if it does not already exist.

faultyserver committed:
Preparing for objects: Transport needs no extra processing.
Since (I'm fairly certain) "ruby_wamp_client" uses it's JSON serializer by default, `@session.call` should automatically serialize `args` using `to_json` on that array. Since Objects now implement a `to_json` method, the conversion should be done fully automatically.

Additionally, Objects can be converted to Hashes by any previous middleware (i.e., ObjectEmbed), and will still be interprable by Transport.

faultyserver committed:
Use Objects instead of hashes as event arguments.
This commit (finally) starts using Object instances for event arguments rather than converting to a hash during `ObjectManager#fire`.

This change allows for a few things:
  - Middlewares can discern argument types using `.class` rather than having to guess (*gasp*). This makes certain actions like object embeddding much simpler, as the middleware can simply iterate arguments and perform actions based on their type.
  - Middlewares do fewer Storage lookups. Since they already have an Object instance, there is no need to look up the instance in `@storage` for first level arguments (nested arguments are still given as identifiers, requiring lookups to get their attributes. This will likely not change).
  - More things that I haven't thought of, but I'm sure exist.

faultyserver committed:
Add `Middleware::default_event_handler` to get/set default handlers.
When called with a block, as `default_event_handler{ |channel, args, kwargs| }`, the given block will be used as the default proc for the `@event_handlers` hash.

When called without a block, as `default_event_handler()`, returns the current value of `@default_event_handler`. Unless explicitly set by a middleware class, this will be the blank proc `Proc.new{}`.

faultyserver committed:
Rename `ObjectEmbed` to `Serializer`.
Per #11, the `ObjectEmbed` middleware is more aptly named `Serializer`, since object embedding has no actual effect until an Object is serialized, meaning the middleware would be serializing objects before doing any embedding.

Classifying the middleware as `Serializer` covers this process semantically, and generalizes the middleware, allowing it to do more than just embed objects (though, it probably won't any time soon).

faultyserver committed:
(#14) Added `Shark::Event` as wrapping class for events
Using the class makes argument passing through middlewares much simpler (one argument vs. two + splat + double splat), and also makes assignment operations on `args` and `kwargs` possible.

faultyserver committed:
(#11) Mostly-working serialization on Object class
The `ObjectSerialization` module includes all of the logic necessary for serializing objects, including configurability through a number of options (outlined in `config/serialization.rb`).

Somewhat awkwardly, `Object` is now indirectly tied to `Storage`, which means that a given Shark instance can only use one Storage instance. This was already the case, but by choice, not by requirement (though, having different adapters for a given instance would be rather odd. Sharding, for example, should be taken care of by the adapter itself, rather than configuration).

faultyserver committed:
`Object#identifier` is now universally-unique.
Calling `.identifier` will always return the full, universally-unique identifier for an Object. This removes the confusion that came with having different local identifiers and universal identifiers.

This change also means that ObjectManagers no longer have a `namespace` attribute, since the universally unique identifier is already namespaced to the Object type (the main purpose of the `namespace` argument).

faultyserver committed:
Add `String#identifier?` to test identifier-ness of strings.
The method returns true if the string could be considered an Object identifier.

faultyserver committed:
Always attempt to serialize every attribute.
When iterating the list of attributes to include in an Object's serialization, the serializer will try to serialize each attribute as it comes up.

For simple types, this means nothing. For containers like Array and Hash, every element will be serialized. For Shark::Object instances, they will be serialized as nested attributes.

faultyserver committed:
Sources that depend on `NormalSource#update` require full identifiers.
Because `NormalSource#update` calls `manager.find_or_new`, sources that rely on that default implementation must supply full identifiers (as would be returned by `Object#identifier`), so that existing objects can be looked up in the storage adapter.

faultyserver committed:
Embedding no longer occurs in Conductor
Since serialization entails embedding, and serialization occurs inside Objects themselves, the embedding processes in Conductor do not need to exist.

Finally, this might be almost done.

faultyserver commented:
I'm ready to merge this, but I haven't tested it through crossbar yet, so I can't effectively analyze the output being generated (I don't have a staging instance set up).

Probably tomorrow (if it doesn't mess you up too much, @elliottwilliams) I'll pause production for a few minutes to look at how this is working.

elliottwilliams commented on middlewares/conductor/station_events.rb:
```diff
@@ -5,31 +5,31 @@ class Conductor
   # Sent when a vehicle departs from this Station. `vehicle` will be the
   # attributes of the vehicle that departed.
-  register_handler 'stations', :depart do |channel, args, kwargs|
+  register_handler 'stations', :depart do |event|
```
Out of curiosity, what makes this empty handler necessary?

faultyserver replied:
I added those just for consistency and clarity.

With these, Conductor has a handler implementation (granted, some empty ones) for every event on the "WAMP Events and RPCs" wiki page.

They could easily be left for `default_event_handler` to take care of, but adding all of them for completeness wasn't much extra trouble.

elliottwilliams commented:
Looks excellent. Feel free to mess with prod. I'll probably be working on Proper tomorrow but can use a local instance.

The merge conflict, btw, was me correcting the same typo you did, so you can easily rebase and drop that commit.

faultyserver commented:
Everything seems to be working. I just need to set `nested_embedded_attributes` for `Shark::Vehicle`, and default `embed_associated_objects` to false instead of true.

Will merge after these changes.

faultyserver committed:
By default, do not embed associated objects.
The functionality should be opt-in, and require specification of which objects should be embedded, rather than which should not.

Opt-in means that most objects will not have to set the configuration option, since they do not need embeds there (it seems like just Routes use the option for Stations, as of now).

faultyserver committed:
Improve specific serializations
- Routes now need to opt-in to embed stations in `associated_objects`.
- When nested, vehicles should only embed a few attributes.
